### DISCUSSION TITLE
Rethink of `Router::state`

### DISCUSSION THREAD

**Discussion by arniu** [tokio-rs/axum#1343] [Ideas]
The state is usually used for data sharing. It can be a db connection, cache client, HTTP client, or a couple of them, which is always wrapped in `Arc`.

**Comment by davidpdrsn** [comment]
I don't understand, what is your question? Please elaborate.

**Comment by arniu** [comment]
https://github.com/tokio-rs/axum/blob/4c9edb4cd4c9d18841bdaf68b67083d1f197dc90/axum/src/routing/mod.rs#L63-L65

Change `state` from `Arc<S>` to `S`.

**Reply by davidpdrsn** [reply]
Your state is still `S`. It's wrapped in a arc internally to ensure it's always cheap to clone. You cannot change that.

You can use https://docs.rs/axum/0.6.0-rc.1/axum/struct.Router.html#method.with_state_arc to use an existing arc, but you're still extracting `S`

**Reply by arniu** [reply]
To wrap the state in an `Arc` or not, just let the user make its own decision. 

The state, usually a db connection, cache client, or a couple of them, is wrapped in `Arc` already. If not, just wrap it manually,  which is cheap and easy.

And as a result, we will make the API much cleaner.

**Reply by davidpdrsn** [reply]
I still think always arc'ing it is the right thing for most users.

How much downside is there really?

**Reply by arniu** [reply]
The downside is:

1. It exposes the implementation details to end users, and the API, related to state, is doubled;
2. In most cases, the state is already wrapped in `Arc`. There is no need to do it again.

**Reply by davidpdrsn** [reply]
I disagree. I think this is the right thing for most users.

**Reply by davidpdrsn** [reply]
To elaborate:

> It exposes the implementation details to end users, and the API, related to state, is doubled;

You mean having `with_state` and `with_state_arc` is "doubled"? I don't think thats an issue. Most people will use `with_state` and the docs for `with_state_arc` explains why it exists.

I guess we're leaking the fact that axum has an `Arc` on the inside but I don't see that as a problem. I doubt it'll change.

> In most cases, the state is already wrapped in Arc. There is no need to do it again.

I'm not sure thats true. Database connections pools sure. But I still think users keep state that aren't necessarily `Arc`'ed, maybe API keys or access tokens.

I don't think two pointer indirections are anything worth worrying about. Your app already does tons of IO so how much does two pointers _really_ matter?

**Reply by arniu** [reply]
What's the downside if the `Arc` was taken off?

**Reply by davidpdrsn** [reply]
Users might use state thats expensive to clone.

**Reply by arniu** [reply]
While it's not expensive to wrap the state with `Arc` manually.

**Comment by mlodato517** [comment]
Swinging over from https://github.com/tokio-rs/axum/issues/1374 I too am confused. I feel like there are a couple ways I would expect `with_state` to work as an end user (since I have no idea about the internals):
1. `with_state` wraps the passed state in an `Arc` and then handlers can extract `State<Arc<S>>` (the router cheaply clones the `Arc<S>` and hands those out to the handlers that need them)
2. `with_state` does nothing to the input and just `Clone`s it and handlers extract `State<S>` and the user should wrap it in an `Arc` for cheaper cloning

It seems that currently an `Arc<S>` is stored internally for cheap cloning but then handlers have to extract `State<S>` which clones the underlying `S` out from under the `Arc<S>` that's stored. If I have that wrong then apologies for being confused yet again :sweat_smile: If I have that right then I think that's a surprising API. I'm just one person but I think some less surprising experiences could be either:
1. Having `with_state(S)` result in `Router<Arc<S>>` so that handlers can extract `State<Arc<S>>`. In this case `with_state_arc(Arc<S>)` also results in `Router<Arc<S>>` with handlers extracting `State<Arc<S>>`.
2. Having `with_state(S)` result in `Router<S>` (as is the case now) but there is no `Arc` and `S` is deep cloned when extracted as `State<S>` (which I think is what's suggested by the original post)

**Comment by arniu** [comment]
曝露内部细节相当于引入额外限制，让我们在设计时缚手缚脚。

而要探求 `state` 长什么样，让我们忘掉具体实现，从需求出发：

1. 保留类型信息，编译时即可发现接口是否匹配；
1. 支持路由嵌套；
1. 便于使用；
1. 易于复制；

这意味着：

- `T1: From<S1>`, 
- `T2: From<S1>`, 
- `T1: From<S1 + S2>`,
- `T2: From<S1 + S2>`,
- `S1: Clone`,
- `S2: Clone`,

> 未完待续

**Reply by davidpdrsn** [reply]
@arniu Please stick to English.

**Comment by ttys3** [comment]
Some frameworks define a `Data` type for `state` like:

```rust
pub struct Data<T: ?Sized>(Arc<T>);

impl<T> Data<T> {
    /// Create new `Data` instance.
    pub fn new(state: T) -> Data<T> {
        Data(Arc::new(state))
    }
}
```

and it impl

```rust
impl<T: ?Sized> Clone for Data<T> {
    fn clone(&self) -> Data<T> {
        Data(Arc::clone(&self.0))
    }
}
```

so that, when extract, it is like `var: Data<T>`

T is not cloned, but Arc<T> is cloned.

**Comment by andersk** [comment]
This seems to have been resolved in 0.6.0-rc3 by #1460.
